src/stack.ts

```diff
@@ -11,172 +11,252 @@
+    public appOptions?: AppOptions;
 
     /**
      * The directory to which cdk synthesizes the CloudAssembly
      * @internal
      */
-    public assemblyDir: string;
+    public assemblyDir!: string;
 
     /**
-     * Any stack options that are supplied by the user
      * @internal
      */
-    public options?: StackOptions;
+    readonly dependencies: CdkConstruct[] = [];
+
+    private readonly createFunc: (scope: App) => AppOutputs | void;
+    private _app?: cdk.App;
+    private appProps?: cdk.AppProps;
+
+    constructor(id: string, createFunc: (scope: App) => void | AppOutputs, props?: AppResourceOptions) {
+        super('cdk:index:App', id, props?.appOptions, props);
+        this.appOptions = props?.appOptions;
+        this.createFunc = createFunc;
+        this.component = this;
+
+        this.name = id;
+        this.appProps = props?.appOptions?.props;
+        const data = this.getData();
+        this.converter = data.then((d) => d.converter);
+
+        // This grabs the outputs off of the stacks themselves after they
+        // have been converted. This allows us to present the outputs property
+        // as a plain value instead of an Output value.
+        const outputs = this.converter.then((converter) => {
+            const stacks = Array.from(converter.stacks.values());
+            return stacks.reduce(
+                (prev, curr) => {
+                    const o: { [outputId: string]: pulumi.Output<any> } = {};
+                    for (const [outputId, args] of Object.entries(curr.stack.outputs ?? {})) {
+                        o[outputId] = curr.processIntrinsics(args.Value);
+                    }
+                    return {
+                        ...prev,
+                        ...o,
+                    };
+                },
+                { ...this.outputs } as pulumi.Output<{ [outputId: string]: pulumi.Output<any> }>,
+            );
+        });
+        this.outputs = pulumi.output(outputs);
+        this.registerOutputs(this.outputs);
+    }
 
     /**
      * @internal
      */
-    public dependencies: CdkConstruct[] = [];
+    public get app(): cdk.App {
+        if (!this._app) {
+            throw new Error('cdk.App has not been created yet');
+        }
+        return this._app!;
+    }
 
-    constructor(public readonly stack: Stack) {
-        super('cdk:index:Stack', stack.node.id, {}, stack.options);
-        this.options = stack.options;
-        this.dependencies.push(stack.pulumiSynthesizer.stagingStack);
+    protected async initialize(props: {
+        name: string;
+        args?: AppOptions;
+        opts?: pulumi.ComponentResourceOptions;
+    }): Promise<AppResource> {
+        const cli = AwsCdkCli.fromCloudAssemblyDirectoryProducer(this);
+        this.appProps = props.args?.props;
+        this.appOptions = props.args;
+        try {
+            // TODO: support lookups https://github.com/pulumi/pulumi-cdk/issues/184
```

Comment:
Is this what will actually populate the `cdk.context.json`?

Reply:
That's right